File: src/utilpc/PP.java
package utilpc;

import java.awt.Dimension;
import java.awt.Point;
import java.awt.Rectangle;
import java.awt.geom.Point2D;
import java.awt.geom.Rectangle2D;

import common.util.system.P;

public strictfp class PP extends P {

	public PP(Dimension d) {
		super(d.getWidth(), d.getHeight());
	}

	public PP(double d, double e) {
		super(d, e);
	}

	public PP(Point p) {
		super(p.getX(), p.getY());
	}

	public PP(Point2D p) {
		super(p.getX(), p.getY());
	}

	@Override
	public PP copy() {
		return new PP(x, y);
	}

	@Override
	public PP divide(P p) {
		x /= p.x;
		y /= p.y;
		return this;
	}

	@Override
	public PP sf(P p) {
		return new PP(p.x - x, p.y - y);
	}

	@Override
	public PP times(double d) {
		x *= d;
		y *= d;
		return this;
	}

	@Override
	public PP times(double hf, double vf) {
		x *= hf;
		y *= vf;
		return this;
	}

	@Override
	public PP times(P p) {
		x *= p.x;
		y *= p.y;
		return this;
	}

	public Dimension toDimension() {
		return new Dimension((int) x, (int) y);
	}

	public Point toPoint() {
		return new Point((int) x, (int) y);
	}

	public Point2D toPoint2D() {
		return new Point2D.Double(x, y);
	}

	public Rectangle toRectangle(int w, int h) {
		return new Rectangle((int) x, (int) y, w, h);
	}

	public Rectangle toRectangle(P p) {
		return new Rectangle((int) x, (int) y, (int) p.x, (int) p.y);
	}

	public Rectangle2D toRectangle2D(P p) {
		return new Rectangle2D.Double(x, y, p.x, p.y);
	}

}

File: src/utilpc/PP.kt
package utilpc

import common.system.P
import java.awt.Dimension
import java.awt.Point
import java.awt.Rectangle
import java.awt.geom.Point2D
import java.awt.geom.Rectangle2D

@Strictfp
class PP : P {
    constructor(d: Dimension) : super(d.getWidth(), d.getHeight()) {}
    constructor(d: Double, e: Double) : super(d, e) {}
    constructor(p: Point) : super(p.getX(), p.getY()) {}
    constructor(p: Point2D) : super(p.x, p.y) {}

    override fun copy(): PP {
        return PP(x, y)
    }

    override fun divide(p: P): PP {
        x /= p.x
        y /= p.y
        return this
    }

    override fun sf(p: P): PP {
        return PP(p.x - x, p.y - y)
    }

    override operator fun times(d: Double): PP {
        x *= d
        y *= d
        return this
    }

    override fun times(hf: Double, vf: Double): PP {
        x *= hf
        y *= vf
        return this
    }

    override operator fun times(p: P): PP {
        x *= p.x
        y *= p.y
        return this
    }

    fun toDimension(): Dimension {
        return Dimension(x as Int, y as Int)
    }

    fun toPoint(): Point {
        return Point(x as Int, y as Int)
    }

    fun toPoint2D(): Point2D {
        return Point2D.Double(x, y)
    }

    fun toRectangle(w: Int, h: Int): Rectangle {
        return Rectangle(x as Int, y as Int, w, h)
    }

    fun toRectangle(p: P): Rectangle {
        return Rectangle(x as Int, y as Int, p.x as Int, p.y as Int)
    }

    fun toRectangle2D(p: P): Rectangle2D {
        return Rectangle2D.Double(x, y, p.x, p.y)
    }
}
